### PR TITLE
Meta Progression

### PR DESCRIPTION
**Pull request by bombasticSlacks**
#### Summary
Features "Meta Progression"

#### Purpose of change
Some scenarios are really complicated or require a lot of game knowledge. These aren't great introductions to the game so they should be locked behind some experience by default. Also gives people a chance to work towards stuff, build linear stories and have their challenge scenarios have easter eggs to unlock further professions. 

#### Describe the solution
Scenarios and Professions now can have a requirement which is an achievement ID

To play said start you need to have completed the achievement on some past character. 

added UI support to explain this in character creation:
* always in the sidebar (so you can see even if you have it unlocked)
* at the bottom if you don't, 
* in a popup if you try to select it anyways. 
* They are also greyed out and sorted to the bottom of the lists by default.

now lab starts require you to actually have cleared a lab finale.

Meta Progression can be enabled per world. 
Meta Progression can be permanently disabled using the debug menu.

Random starts respect what is unlocked.
Preset characters **Do Not** rationale being any time you are using a preset it should work out of the box.

#### Describe alternatives you've considered
Don't over engineer this as much, I really went overboard with the infrastructure on this one.

Add more locked scenarios but for now I'm gonna hold off and keep them to separate PRs where they could be argued individually. 

Meta Progression could be a proper option however for example I like to have a debug world (where this would be disabled) and a "standard" world where I actually play.

In the future additionally options could be added to only track meta progression for the world you are currently playing allowing for challenge runs on specific mods.

#### Testing
Loads up fine in the character creator.

#### Additional context
Example Profession (I didn't actually lock this one behind impenetrable this is **just** an example)
image

Labs being locked
image

If you've completed the achievement
image

World Gen Option
image

Debug Override
image
image

### CODE REVIEW

**Comment by TheShadowFerret** [comment]
Can this be further specified to ban locations w/o an achievement(ex. can't start at refugee center w/ Safe Place without completing The Last Homely House/finding it)?

**Comment by sonphantrung** [comment]
How will this play out to Experienced players where they have done a lot of challenge runs?

**Comment by bombasticSlacks** [comment]
If you mean what happens to current players and will they now be locked out of content they've already done the answer is no. The game already tracks past achievements so if someone has cleared a lab it already works.

**Comment by Night-Pryanik** [comment]
I don't like this at all, for several reasons.

First of all, I hate it when content is locked by requirement of previous runs. Especially in sandbox games. I'm a total newbie and I want to play ultra-hardcore challenge? That's completely my business and no one should prevent me from selecting it, even if it will result in my character's death.

Second, what if experienced player, who got all the achievements in his previous runs, deleted the game some time ago, and now is returned from hiatus? He must complete achievements once again to get challenge scenarios?
The same for scenario where game folder was deleted for some other reasons, e.g. disc failure.

And third, as developer, how do I fix the bugs with some challenge scenarios if I don't have required achievements?

All in all, this feature should be made togglable.

**Comment by bombasticSlacks** [comment]
I was planning on adding a toggle I'm just not sure if it should be a 

game option
A debug option to unlock all achievements.
A world option. 

@kevingranade had some opinions when we last talked about it but the mechanism for checking achievements is much simpler now than we first discussed.

**Comment by Night-Pryanik** [comment]
Personally I'd prefer it to be a debug option, but without linking to achievements. I might want to get achievements in challenge scenarios right from the start. Simple "Unlock all professions/scenarios" would suffice.

**Comment by PatrikLundell** [comment]
I dislike it for the same reasons as @Night-Pryanik. I "upgrade" by installing the latest experimental as a new folder to avoid removed files remaining and messing things up, and then copy my current save over. When I start a new game I generate a new world (which may be in a new install). Also, whenever debug hacks are used achievements are disabled, which I don't care about as such (I know achievements can be re-enabled, but I just don't care enough to bother).

Also, I strongly dislike games that lock away contents behind requirements to have played previous sessions, typically to "extend game play".

I believe challenge scenarios should be described as such, and if you want to disregard the warning that should be there it's your business. The same goes with other "extra challenge" options.

Hiding "challenge" options from players by requiring them to find how to enable them via debug options (that are themselves hidden until you enable them) may be useful for developers and sufficiently experienced players, but not for regular players who would think debug options are for debugging. If you insist on hiding such contents it should be through a character generation option that would be disabled unless the player enables it (similar to how you select the point system to use). Whether it should be a single "challenge" switch or more granular per category is a design detail.

**Comment by bombasticSlacks** [comment]
I think you all are entitled to your passion when describing how this might be a mild inconvenience for power users.

As for new players who just need to "suffer through it" or "git gud" This game isn't Kenshi, the scenarios aren't on a sliding scale of hard to easy with the only reward being bragging rights each scenario gives points. So you start up the game and see that lab starts give 8 points or starting as a cyborg gives 4 (not bothering to read the "broken" part). You pick them so you get more Skillz, start playing and start dying for no clear reason. Or are in a room you don't understand how to get out of, youre not fighting a hard boss, just unable to leave a room. This kind of stuff makes people assume the free open source game is broken and bounce off the game entirely.

So this isn't supposed to be some long term 100 hours to unlock everything grind. Just a way to help ease new players into the game and the brainstorming on it started because of repeatedly helping new players on the discord with complex game systems who don't actually understand any of the simple systems yet.

**Comment by Night-Pryanik** [comment]
I'm not sure why should I tolerate this kind of inconvenience and limitation. For what? In order not to hurt feelings of some random dumbass who can't read things right?

**Comment by Inglonias** [comment]
> I don't like this at all, for several reasons.
> 
> First of all, I hate it when content is locked by requirement of previous runs. Especially in sandbox games. I'm a total newbie and I want to play ultra-hardcore challenge? That's completely my business and no one should prevent me from selecting it, even if it will result in my character's death.
> 
> Second, what if experienced player, who got all the achievements in his previous runs, deleted the game some time ago, and now is returned from hiatus? He must complete achievements once again to get challenge scenarios? The same for scenario where game folder was deleted for some other reasons, e.g. disc failure.
> 
> And third, as developer, how do I fix the bugs with some challenge scenarios if I don't have required achievements?
> 
> All in all, this feature should be made togglable.

I agree entirely. The first thing I can think of is that as a developer, I'm rebuilding the install in my local repo all the time, and more importantly, don't play in it regularly. I don't have the achievements there.

I also agree with the sandbox philosophy. Putting "Challenge" in front of these scenarios makes it clear to new players that they shouldn't seriously attempt them unless they have some idea of how to play because they are, well, challenging.

Making this feature optional or toggleable completely defeats the purpose of including it as well. If you can turn something like this off easily, it may as well not be there.

**Comment by PatrikLundell** [comment]
I again agree with @Night-Pryanik. I don't consider it to be unreasonable to require that people read the description before greedily taking all those "free" point. If "get gud" means acquiring a basic reading ability for a game that contains a lot of reading, then I'm for it.

However, that also means scenarios should be clearly labeled, and it wouldn't hurt to have an overall description informing players that things are worth points because they add various difficulties, and those looking for free points might want to look at the free point setting (or whatever it's called). I don't remember if the labeling currently is as clear as it should be, though.

I do not fully agree with @Inglonias, as cutting away a lot of chaff you don't want to have to wade through and don't want to use has its own benefits. The game has lots of useful filters for this kind of thing, but I don't think there's an easy way to make a filter that's easy to use without prior knowledge of what to filter for.

**Comment by sonphantrung** [comment]
Just for some references: the Hitman World of Assassination trilogy games (1/2/3) actually disallows you to start in some locations and disguises unless you met the required mastery level

**Comment by PatrikLundell** [comment]
And that's a crap implementation. A lot of games have unlocks to extend play time because it's good for their metrics, not because it's good for their players.

**Comment by sonphantrung** [comment]
> And that's a crap implementation. A lot of games have unlocks to extend play time because it's good for their metrics, not because it's good for their players.

Definitely. It's just for some "references", one can see it 'crappy', one can see it 'good', based on his opinions

**Comment by liminalRoar** [comment]
I wouldn't see why having a debug option for developers is not a good solution.

I personally like the idea of unlocking certain options through meta progression.

I really love the idea of a custom profession option someday.  Ability to choose your own gear and everything with a set amount of points to spend (gear could even be unlockable through playthroughs as well for custom characters. Could possibly be it's own mode or an unlockable option with some achievement.

**Comment by fairyarmadillo** [comment]
This is an annoying implementation that conflicts with the gonanywhere do anything design at the core of the game, and should not be merged. Difficult scenarios are already self-limited by player knowledge, and artificially locking them behind arbitrary acheivements is just a nonsensical speed bump. The game has never and should not work this way.

**Comment by fairyarmadillo** [comment]
Being rude to people is not really helping your case.

File: src/newcharacter.cpp [hunk 1]
```diff
@@ -3068,11 +3129,16 @@ tab_direction set_scenario( avatar &u, pool_type pool )
                        "                                             " );
             nc_color col;
             if( get_scenario() != sorted_scens[i] ) {
+                if( sorted_scens[i]->has_flag( "CITY_START" ) && !scenario_sorter.cities_enabled ) {
+
```
[the remark this answers is not in the document]
Comment (bombasticSlacks) [reply src/newcharacter.cpp:3133]: Just a mistake while copying stuff to appease clang, I'm surprised it passed clang again thanks

**Comment by kevingranade** [comment]
For experienced players and developers, I favor a debug menu option to "unlock all scenarios" implemented as a fake achievement that (ideally) doesn't appear in any achievement listings.  IIRC you can set debug menu entries to either disable achievements or not, this is one that should not do so, it doesn't make getting other achievements particularly easier so it doesn't invalidate them.
That also means you should just be able to copy your global achievements file around to new installs, which will pull your old achievements along, including the fake "unlock everything" achievement.

**Comment by Inglonias** [comment]
Is there a good reason to have this when devs can turn this off with the press of a button anyway, and less experienced players may submit false bug reports along the lines of "I can't do lab challenges anymore?"

**Comment by anothersimulacrum** [comment]
I doubt anyone will make those bug reports, it's clearly communicated why.

**Comment by I-am-Erk** [comment]
> Is there a good reason to have this when devs can turn this off with the press of a button anyway, and less experienced players may submit false bug reports along the lines of "I can't do lab challenges anymore?"

Yes. Unlocking things by progression is a popular and generally enjoyable gameplay device. There's a very good reason it appears in a lot of games. It also allows us to have a wider design space in late game starts and things than we otherwise would. When the only argument against it is "I am concerned I may have to copy one file I don't normally copy in order to maintain my workflow" I think it's a pretty strong feature proposal.

**Comment by fairyarmadillo** [comment]
This isn't really a feature though, it's just an impediment with no upside.

**Comment by I-am-Erk** [comment]
> it's just an impediment with no upside

That's just, like, your opinion, man. It's a feature *I* like, and would like to play with. I am in fact quite excited about it and have been suggesting it since we got achievements.

>  The fact that it's already got eight thumbs down should tell you something, and the takeaway shouldn't be "oh, those people just don't get it."

Sure, it tells me that any new feature that has any major impact on the game receives a ton of controversy. It also has a bunch of party votes, and like I said, I am personally pretty keen about it. Further, given some of the responses - like yours - I think it's pretty fair to say that some people just "don't get it". If you don't want this option, turn off unlocks on your world and go ham.

> the feature is "popular" elsewhere because commercial titles use it to drive engagement and boost sales. 

Having a career mode in a game like, say, Hardspace: Shipbreaker or Cooking Simulator does nothing to "boost sales and drive engagement". It is *fun*. We are playing a game, and progression is enjoyable in games. However, in both those games and in CDDA once this is merged, there is also the easily accessible gameplay option to go freestyle and ignore the unlock system, thus allowing people to enjoy a sandbox mode with all the options, or a constrained mode where they are mechanically rewarded for pursuing options.

As I said: your argument is that I should not be able to play the gameplay style I enjoy most, one that conveniently also makes the game *much* more accessible to new players, and opens up an entire interesting area of new design content. My argument is that we should both be easily able to play whichever style we like. I do not think you are approaching this in a particularly free or fair manner, when one thinks of it that way.

**Comment by Inglonias** [comment]
My biggest concern was that this feature wouldn't get developer support after the fact - That someone would maybe hide one or two scenarios behind existing achievements and call it a day forever. But from the way erk and slacks is arguing for the feature, it sounds like they're enthusiastic enough to give this a little bit of thought. Basically, if you're going to do this, do it right, and I think you've demonstrated your enthusiasm if nothing else.

**Comment by keampe** [comment]
I'm gonna weigh in on team no, here - not that it'll make a difference. I really dislike it when games block content I want to play behind arbitrary walls of stuff I'm not interested in. And to do this with achievements? I _hate_ achievements, which means I ignore them and really don't care if they're on or not. I don't believe I'm alone on this either. 

Also, which scenarios and which achievements? If you don't do all the challenge scenarios then people will just migrate to the ones that are left. So which achievement do you need to do the Portal Dependent scenario, for example?

So, ya, not a change I'm in favor of. Besides, there are better ways of doing this. I'll give you two off the top of my head in the next comments.

   - Shane

**Comment by keampe** [comment]
First, color code the scenarios. Yep, just make the hard ones bright red and the easy ones green. Then add a single word to the title as to why (apparently the word "Challenge" isn't enough) so - Difficult!!! or Complex!!! You can then add more in the description if you like. Anyone who misses that really has only themselves to blame.

 - Shane

**Comment by keampe** [comment]
Second one might be controversial.

Take away the bonus points.

WHAT!!!!! Fuck You!!! Take away my bonus points - What are you NUTS!!!

Seriously, though. Why do you play a Challenge Game?
 A. For the challenge - it's right in the description!
 B. For the bragging rights - Yup, I did that and I started NAKED!
 C. To make the uberpowerfulgod character.

If it's A or B, the fact that you got extra points for this sorta mitigates it, right? I mean winning the national Belgian Tennis Tournament is quite an achievement worthy of bragging right, right? Unless you're Roger Federer. Then it's like Ya, so what, what're you doing here anyways - you're not Belgian.

If it's C, then we can always add a Scenario called "Training Wheels" that is just like the standard start but gives you 8 extra points. We can also make a career called "Total Noob" that starts you with a suppressed Glock and a Fire Axe, if you like. Or you can just change your point total. Or use free form.

Anyway, the points from actual disadvantages should probably stay, but everything else, why is it there in the first place?

   - Shane

**Comment by I-am-Erk** [comment]
That is already a planned change, and doesn't really resolve the core point of "lots of us like playing with progression, and those that don't will still have the option to play as it was".

**Comment by Qrox** [comment]
Some questions: how does this interact with the "random character/play now" and "preset character" starts? Does it randomize from all professions/scenarios or only unlocked ones? Does it reject preset with locked professions/scenarios or does it just accept the preset?

Also I feel an option in the character creation menu to toggle meta progression can better communicate the fact that this feature is optional, since many people don't usually fiddle with game options. This can be done similar to how some video games have "career" and "free" modes, although we probably need different terms for them and the "free" mode should come with a warning that the "career" mode is recommended for new players for them to familarize themselves with the game.

**Comment by scarf005** [comment]
`customer`: A cup of espresso please.
`barista`: Well, from today we're introducing 'progressions' to our menus. Which means you must purchase 10 americanos first to unlock espresso. And americano is also locked, you require 20 cafe latte purchases to unlock that.
`customer`: What? just give me espresso.
`barista`: You see, many customers drink espresso not knowing it's very bitter, having bad experience. So from now on we're encouraging our customers to be able to handle more bitter drinks with this progression system.
`customer`:  I've been drinking espresso from other cafes, I *know* espresso is bitter. Give me the espresso.
`bariasta`: That's from other cafes. You've only purchased 5 cafe lattes here.
`customer`: I doubt everyone would hate that.
`barista`: Well, that's your opinion, and I'm very excited with this system. Progression is generally a enjoyed feature.
`customer`: How come limiting what customer can choose be called 'progression' and be condiered enjoyable?
`barista`: Well if you *insist* not to follow our system, you could dial our ARS to bypass the block.

I'm noticing some similarities...

**Comment by Inglonias** [comment]
@Qrox a smoother solution might be to disable the progression locks if a user chooses the freeform points pool, to mark it as a sandbox mode.

**Comment by bombasticSlacks** [comment]
> Some questions: how does this interact with the "random character/play now" and "preset character" starts? Does it randomize from all professions/scenarios or only unlocked ones? Does it reject preset with locked professions/scenarios or does it just accept the preset?

The next thing Im gonna do is make it so you can flag any option as disabled for random runs.  I personally play random and find a lot of the challenge runs not very fun for all random.  Preset I'm not sure I'll test that later today probably. 

>  better communicate the fact that this feature is optional, since many people don't usually fiddle with game options

My thought would be if the player is intelligent enough to understand the difficulty of challenge runs and make an informed decision of diving right into them they are also capable of reading the world settings. If players don't want their hand held when it comes to picking starts then Im not gonna hold their hand by making the setting more obvious than a clearly displayed toggle. 

> solution might be to disable the progression locks if a user chooses the freeform points pool, to mark it as a sandbox mode.

This was my first thought but the issue is all the proposed overhauls to character creation have us moving towards free form with descriptives. I'll edit this to link the Issue in a few.

**Comment by I-am-Erk** [comment]
Scarf, for the amount of time you took to type that analogy, you could manually turn the progression option off in every single world you'll ever make of cdda. I can't believe the amount of complaining here to a problem that can be solved by turning a single readily-accessible ticky box to "no"... Assuming you don't want to just go into debug options and permanently disable it.

**Comment by Ilurker** [comment]
Just skimming the comments here. There are clearly some strong opinions. I very solidly fall in the camp that putting some newbie bumpers on games helps more players enjoy and stick with the experience than it drives skilled players away, and I support the PR as it stands.

I've had the experience of writing literal pages of help for new players who picked the cyborg start because it sounded really cool, only to get very mad at the game when it was a world of pain. CDDA already has a brutal learning curve for players and just about all of the coolest sounding starts - even the ones that confer an advantage - turn that into more of a cliff. This gets especially bad when you regularly see people talking about how lab and TCL are the best starts, both of which will generally take a new player and slam them into the pavement eight or nine times before they figure out the basic control schemes. To someone who knows the game, it might be obvious that they just took a bad newbie pick and should try something else, but to a new player, nothing is obvious yet, and we're gambling that they won't abandon a new (free, so no sunk cost) game rather than tough it out.

If it hasn't been suggested before, couldn't much of the philosophical debate over whether it's alright to lock progress be solved with an options switch disabling progression gating? I'd imagine a dialogue window that pops up during initial character generation with a short paragraph explaining that some scenarios are hidden to both save surprises for a new player, and warning that it's very possible to create a character or take a scenario for whom the game will be more or less impossible for anything but the most veteran of players.

Then, give them the option to disable achievement gating. Default the initial selection to the one that suggests you keep some scenarios locked to present keymashing to to a failure state, and hide the toggle in the options menu in case a veteran player keymashes into a lockout. At that point, the newbie lock will very much be a newbie lock, and for most players will only be served once. For those of us using odd upgrade methods, a single extra dialogue on the first character gen only really isn't much of a burden.

**Comment by Ilurker** [comment]
Sorry to double comment, but if the intent is really to make achievement gating a core part of the gameplay loop, perhaps make the lockout optional as a debug override? Same idea, just a less accessible menu that *really* selects for veterans, and has its own warning already attached?

**Comment by bombasticSlacks** [comment]
> Sorry to double comment, but if the intent is really to make achievement gating a core part of the gameplay loop, perhaps make the lockout optional as a debug override? Same idea, just a less accessible menu that really selects for veterans, and has its own warning already attached?

Way ahead of you, check the images in the PR

**Comment by Ilurker** [comment]
> Way ahead of you, check the images in the PR

Of course I missed that. 🤦‍♂️

**Comment by scarf005** [comment]
> I can't believe the amount of complaining here to a problem that can be solved by turning a single readily-accessible ticky box to "no"...

Maybe *the actual problem* is that cataclysm still lacks proper tutorials. Heard of [XY problem?](https://en.wikipedia.org/wiki/XY_problem) Making longer, better tutorials would probably be better for new players?

**Comment by bombasticSlacks** [comment]
If you would like to make more tutorials they would be welcomed.

**Comment by I-am-Erk** [comment]
> Making longer, better tutorials would probably be better for new players?

Sure, go for it. That still misses a fundamental point that is common to the complaints in this thread: it wouldn't provide anything for those of us who are quite excited to play in achievement unlock mode. Since that's the main reason I personally am excited for this feature, I don't think tutorials are the "actual problem".

**Comment by sonphantrung** [comment]
This reminds me a lot about the SystemD being implemented into Linux Distros: Many veteran Linux users just hate SystemD

**Comment by scarf005** [comment]
why did you delete my comment? my comment was neither offensive nor violating rules of github community. I thought CDDA had freedom of speech? what are you fearing?

**Comment by I-am-Erk** [comment]
Scarf, go to Reddit or discourse or something if you want to make a bunch of off topic posts. This is a pill request, not a chat forum.

**Review by NetSysFire — COMMENTED** [review]
There are pretty mixed opinions on this and I *generally* see this as a good thing.

Also, I wonder if you should use terminal punctuation in the marked strings since it appears that everything else also uses it and that would introduce inconsistency.

Another thing is the capitalization of "Arcade Mode". This is how it is in the achievement but e.g the y/n query says "arcade mode".

**Comment by DamienRoyan** [comment]
> > I can't believe the amount of complaining here to a problem that can be solved by turning a single readily-accessible ticky box to "no"...
> 
> Maybe _the actual problem_ is that cataclysm still lacks proper tutorials. Heard of [XY problem?](https://en.wikipedia.org/wiki/XY_problem) Making longer, better tutorials would probably be better for new players?

The current tutorial is quite sufficient. I went through it recently so I could walk my friend through the new tutorial (He got stuck trying to figure out how to open a door). Unless the player accidentally skips a dialogue explaining something, the tutorial gives a full rundown of control and environment interaction basics.

**Comment by NetSysFire** [comment]
There *are* issues with the tutorial but this is not the place for that. There is #57571 for issues with the tutorial and general new player experience.

**Comment by silencecoder** [comment]
> a fundamental point that is common to the complaints in this thread: it wouldn't provide anything for those of us who are quite excited to play in achievement unlock mode. Since that's the main reason I personally am excited for this feature, I don't think tutorials are the "actual problem".

The locked content is. CDDA still doesn't have a good on-boarding process for new players. They should be guided into a proper entry point (an introduction scenario) or provided with an embedded tutorial in each scenario instead of being restricted in their choices. If a new player want to start as a cyborg in labs, then the game should warn player about the upcoming hardships but ultimately leave the decision to a player.

If seasoned players with a sufficient internal progression want an external meta-progression then it should be provided in form of a _campaign_. That way the game would allow a tailored per-player progression in a dedicated mode with a unified list of achievements. Developers can use this list as a dependency for new modular content (easter eggs, special events, characters, plots and more). That way content can be provided 'as is' in a normal play-through or silently toggled in a campaign.

New players also can choose to play a campaign if they prefer a linear storytelling. But this might come at a cost of wrapping their heads around achievements. If new content can be obtained only with unlocked achievement, player need to get a rough idea on available achievements and what they are about. Which in turn may disclose mechanics, events and other content hidden in the game. This can be solved with a curated tiers of achievement, but that approach won't do much for seasoned players other than further narrowing their progression options.

Again, the PR starts with "Some scenarios are really complicated or require a lot of game knowledge. These aren't great introductions to the game so they should be locked behind some experience by default."

**Comment by I-am-Erk** [comment]
A closed thread is not the right place to discuss questions and suggestions for improving tutorials. There was an issue linked above.

**Pull request merged**